### ISSUE TITLE
Encrypt mail with multiple certificate

### ISSUE THREAD

**Issue by nathinthesky** [bbottema/simple-java-mail#297]
hello,

more a question for a new feature than an issue. Is it planned to implement encrypted mail for multiple recipients ?

Thanks !

**Comment by bbottema** [comment]
I'm knowledge on this is limited, but I think it should already work with multiple recipients, if their keys were signed with a shared CA certificate. Or something like that.

**Comment by nathinthesky** [comment]
Sorry for delay to answer.
To encrypt mail for multiple recipients, you have to encrypt it with each recipient certificate (public key), then, they will be able to decrypt the mail on their side with their private key.
I'll will try to work with you on this !

**Comment by bbottema** [comment]
At one of the companies I worked, they configured a single CA cert, which was used to sign/encrypt? all emails with using recipients individual public keys. Is your use case actually really a common one? Seems like a hassle!

**Comment by nathinthesky** [comment]
I think my case is a common one.

Usually, a company has its own CA, they use it to sign internal certificates. Each employee then has a key pair (private and public key).

Let's say I want to send an encrypted mail to Alice and I want her to be the only one to be able to decrypt it. I encrypt it with her public key. The only person able to decrypt the sent mail is the one with the corresponding private key (=Alice).

Now, I want to send an encrypted mail to Alice and Bob (at the same time, = multiple recipients). If I encrypt it only with Alice public key, only Alice will be able to read it. Bob can't. That's why I need to encrypt it twice : one time with Alice's public key, one time with Bob's one.

Right now, if I want to do that with simple-java-mail, I have to send 2 distincts emails instead of one.

I don't know the company you're talking about but if you use only one certificate (you usually don't use a CA for that) to encrypt each email, that means, everybody share the private key and everybody can decrypt them.

The same if everybody use the same key to sign each email, you can't be sure of who really send it. Usually, you sign it with your private key. 

But maybe that's what they intended to do...

By the way, in the lib, sign case is ok because you always sign a mail once (= with one private key).

**Comment by bbottema** [comment]
> By the way, in the lib, sign case is ok because you always sign a mail once (= with one private key).

Ahh then that might have been the case at the company I worked for at the time.

You made a convincing case, one question: this change would mean being able to sign on a per-email basis (so using EmailBuilder). Would it ever make sense to be able to do it on Mailer level as well, affecting all emails that don't have their own key added for signing/encrypting?

**Comment by nathinthesky** [comment]
you mean to encrypt on a per-email basis ?

Not sure to understand your question: "_do it on Mailer level... emails that don't have their own key added for signing/encrypting_"

As encryption is for a specific recipient, I think certificates has to be done explicitely set in emailBuilder (no default value). For signing, it can be configured as default (as you can choose to sign all your email).

**Comment by bbottema** [comment]
>I think certificates has to be done explicitely set in emailBuilder (no default value)

So no use case whatsoever to allow a default. Ever. (just double checking)

**Comment by nathinthesky** [comment]
Sign -> can set a default to sign email
Encrypt -> no default value (as it is recipient dependant)

**Comment by rover886** [comment]
Yes, I am also having the same use case as @nathinthesky explained, encrypting message for each recipient. I just ported my code from Commons Apache to Simple Java Mail and come to know that it is not possible 😞 

With Commons Apache we used to encrypt mail following this flow chart:

1. Fetch each recipient certificate (X509Certificate) from LDAP
2. By using list of X509Certificate and `SMIMEEnvelopedGenerator` from `org.bouncycastle.mail.smime`, `MimeBodyPart` is encrypted

Here is the code for encrypting `MimeMessage` object

    public MimeMessage encryptMessage(MimeMessage message, List<X509Certificate> recipientCertList) 
    {
		
		try {
			Security.addProvider(new BouncyCastleProvider());
			
			// Create the generator for creating an smime/encrypted message
			SMIMEEnvelopedGenerator  generator = new SMIMEEnvelopedGenerator();
			
			//Add Recipient Public Certificates.
			for (X509Certificate cert : recipientCertList) {
				generator.addRecipientInfoGenerator(new JceKeyTransRecipientInfoGenerator(cert)
						.setProvider(BouncyCastleProvider.PROVIDER_NAME));
			}

			MimeBodyPart msg = new MimeBodyPart();
			msg.setContent(message.getContent(), message.getContentType());
			
			MimeBodyPart mp = generator.generate(msg, new JceCMSContentEncryptorBuilder(CMSAlgorithm.AES256_CBC)
					.setProvider(BouncyCastleProvider.PROVIDER_NAME).build());
			message.setContent(mp.getContent(), mp.getContentType());
			message.saveChanges();
			
			return message;
		} catch (Exception e) {
			throw new EmailGenericException(e);
		}
		
		
	}

After this function just use `Transport.send();` with returned message object.  And if messages signing is also need to be done then that has to be done first then encryption. Moreover if DKIM signing should also be the last step. So the flow will always be S/MIME sign --> S/MIME encryption --> DKIM sign.

@bbottema **Really looking forward to have this facility to encrypt message per recipient into this library.**  Also, if I get any rough idea where would the most of the changes will come then I can create PR for the same.

**Comment by bbottema** [comment]
I had no idea that was even possible with LDAP. That's pretty cool actually. As I'm winding down on the open issues, I am looking into some of the older reports, such as this one, so I think I'm going to start work on this soon.

> And if messages signing is also need to be done then that has to be done first then encryption. Moreover if DKIM signing should also be the last step. So the flow will always be S/MIME sign --> S/MIME encryption --> DKIM sign.

Correct, this is how Simple Java Mail already has implemented this.

What I'm going to do is to allow both default signing or encrypting on Mailer level as well as on Email level and then everyone can figure out there own use case and use the library as they see fit.

**Comment by jgprospect** [comment]
I also would like to send an encrypted email to multiple recipients that is also signed. Any idea if/when this might make it in?

**Comment by jgprospect** [comment]
bbottema, the documentation you reference to RecipientBuilder looks great. I have version 8.12.4 but I don't see RecipientBuilder in the jar. I also have the SMIME module and the core module and I don't see it listed anywhere. Am I looking in the wrong place? How should I get this functionality? Thanks!

**Comment by bbottema** [comment]
I think I've made a mistake with the documentation, it should have been in a separate (unreleased) branch. Sorry for the confusion.

I intend to pick this up again, but I'm currently on a break from the library. I've been developing it for almost 19 years now, and every now and then I take a long break (6-12 months) before diving back in. Unfortunately, I can't tell when I'm going to pick it up again, sorry. I have a lot going on right now. 

I do still pick up security patches.

**Comment by bbottema** [comment]
Fixed by making S/MIME encryption certificate selection recipient-aware.

A `Recipient` can now carry its own `X509Certificate` via `RecipientBuilder.withSmimeCertificate(...)`. When any TO/CC/BCC recipient has a certificate, Simple Java Mail uses the per-recipient encryption path; recipient-level certificates override the email-level default, while the email-level `SmimeEncryptionConfig` can still provide fallback certificates and algorithms.

Example:

```java
Recipient alice = new RecipientBuilder()
    .withAddress("alice@example.com")
    .withType(Message.RecipientType.TO)
    .withSmimeCertificate(aliceCertificate)
    .build();

Email email = EmailBuilder.startingBlank()
    .from("me@example.com")
    .to(alice)
    .withPlainText("Encrypted for Alice")
    .buildEmail();
```

For multiple recipients, build each `Recipient` with its own certificate and pass them through `.to(...)`, `.cc(...)`, or `.bcc(...)`.

**Comment by bbottema** [comment]
Released in Simple Java Mail v9.0.0, which is now available from Maven Central:

https://repo1.maven.org/maven2/org/simplejavamail/simple-java-mail/9.0.0/

Release notes:
https://github.com/bbottema/simple-java-mail/releases/tag/9.0.0
